internal/server/httpapi/trigger.go

```diff
@@ -0,0 +1,398 @@
+									http.StatusInternalServerError)
+								return
+							case *pb.GetJobStreamResponse_Terminal_:
+								// We got some job output! Craft a message to be sent back
+
+								for _, ev := range event.Terminal.Events {
+									log.Trace("job terminal output", "event", ev)
+
+									switch ev := ev.Event.(type) {
+									case *pb.GetJobStreamResponse_Terminal_Event_Line_:
+										value = ev.Line.Msg
+										valueType = "TerminalEventLine"
+									case *pb.GetJobStreamResponse_Terminal_Event_NamedValues_:
+										var values []terminal.NamedValue
+
+										for _, tnv := range ev.NamedValues.Values {
+											values = append(values, terminal.NamedValue{
+												Name:  tnv.Name,
+												Value: tnv.Value,
+											})
+										}
+
+										value = values
+										valueType = "TerminalEventNamedValues"
+									case *pb.GetJobStreamResponse_Terminal_Event_Status_:
+										value = ev.Status.Msg
+										valueType = "TerminalEventStatus"
+									case *pb.GetJobStreamResponse_Terminal_Event_Raw_:
+										value = string(ev.Raw.Data[:])
+										valueType = "TerminalEventRaw"
+									case *pb.GetJobStreamResponse_Terminal_Event_Table_:
+										tbl := terminal.NewTable(ev.Table.Headers...)
+
+										for _, row := range ev.Table.Rows {
+											var trow []terminal.TableEntry
+
+											for _, ent := range row.Entries {
+												trow = append(trow, terminal.TableEntry{
+													Value: ent.Value,
+													Color: ent.Color,
+												})
+											}
+										}
+
+										value = tbl
+										valueType = "TerminalEventTable"
+									case *pb.GetJobStreamResponse_Terminal_Event_Step_:
+										m := ev.Step.Msg
+										if len(ev.Step.Output) > 0 {
+											m = m + "\n" + string(ev.Step.Output[:])
+										}
+										value = m
+										valueType = "TerminalEventStep"
+									default:
+										log.Error("Unknown terminal event seen", "type", hclog.Fmt("%T", ev))
+									}
+								}
+							default:
+								log.Warn("unknown stream event", "event", resp.Event)
+							}
+
+							// Send a message job stream back to the client
+							if value != nil {
+								log.Trace("sending job data to client for job", "job_id", jId)
+								msg := Message{
+									JobId:     jId,
+									ExitCode:  exitCode, // will only be sent if set to non-empty string
+									Value:     value,
+									ValueType: valueType,
+								}
+
+								// Lock to ensure multiple routines don't send back a message at the same
+								// time to the receiver and mess up the incoming message
+								mu.Lock()
+								// send the message back
+								err := enc.Encode(msg)
+								if err != nil {
+									log.Error("failed to encode job stream output to send back", "err", err)
+									http.Error(w, fmt.Sprintf("server failed to encode job stream output: %s", err), 500)
+									return
```

Comment:
Do we need to call `mu.Unlock()` here before we return?

Reply:
Hm, yes I think so!